Cleanup and do not list current user shares in getShares too

skjnldsv opened the pull request:
- getSharesInDir was properly excluding shares where shared_with was the current user (https://github.com/nextcloud/server/pull/14605) but we were missing the check in getShares too
- Added more comments on the code
- Cleanup and move duplicate code into one function

I could use some help on the tests :)

skjnldsv committed:
Unify code to get shares of a file or also including subfiles

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Filter out duplicated shares before processing them

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented on apps/files_sharing/lib/Controller/ShareAPIController.php:
```diff
@@ -1208,4 +1166,49 @@ class_exists('\OCA\Circles\Api\v1\Circles')) {
+		}
+
+		// TALK SHARES
+		$roomShares = $this->shareManager->getSharesBy($this->currentUser, Share::SHARE_TYPE_ROOM, $path, $reshares, -1, 0);
```
@danxuliu do we have a specific shareprovider for talk as well?
So we could use a check like `$this->shareManager->shareProviderExists(Share::SHARE_TYPE_ROOM)` ?

danxuliu replied:
Yes, [it is in Talk](https://github.com/nextcloud/spreed/blob/e9335698fb2fc2ae248a25ec1bfe49dd94bc4abb/lib/Share/RoomShareProvider.php), and [you can use that check without problems](https://github.com/nextcloud/server/blob/e930a0ccbef07f5ce9847c39584d4125aecffe38/apps/files_sharing/lib/Controller/ShareesAPIController.php#L183) ;-)

skjnldsv commented:
All good, please review! :)

rullzer commented:
Master is no 18 development. If this needs to go into 17 please follow the normal backport procedures after merging.

skjnldsv commented:
Same here everyone, please let's get this in so I can continue the sharing revamp!

rullzer reviewed (approved):
Lets do this then

ChristophWurst reviewed (approved):
:elephant:

danxuliu reviewed (changes requested):
There are several failing tests, not ready to release yet. I will take a look.

danxuliu committed:
Replace fully qualified name with alias

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add integration tests for getting shares including subfiles

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add unit tests for getting shares including subfiles

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add unit tests for getting shares of a file

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Remove unneeded calls to "ShareManager::shareProviderExists()"

"ShareManager::getSharesBy()" already checks if the share provider
exists before returning the shares and, if the provider does not exist,
it returns an empty array. Therefore it is not needed to explicitly
check if the provider exists or not.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Unify share type sorting when getting shares of a file or with subfiles

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add missing share types when getting shares of a file

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Extract method to get all shares

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Do not return shares with the current user

Getting the shares of a file no longer returns shares with the current
user for consistency with the results when getting the shares including
subfiles.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add more integration tests for getting shares

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu commented:
Sorry, it took me way longer than expected :-(

I have added more unit and integration tests, split the fixes and the refactorings, did some minor cleanups and fixed the pending issues (I hope :-P ).

@skjnldsv I moved [`if (!$resharingRight && $this->shareProviderResharingRights($this->currentUser, $share, $path)) {`](https://github.com/nextcloud/server/commit/bf6209c24422e07f2a185f1836c1a5d9b59a0d80#diff-dbbf5139c4d76614e3af7d2b407c4946R709) back to its original position because in the previous code it was called after `if ($share->getSharedWith() === $this->currentUser) {`, and given that `getShareWith()` is checked too in `shareProviderResharingRights()` that might change the behaviour (although no test was affected, and I was not able to make a test that caused a behaviour change... so if you changed it for some specific reason please provide a test that shows it :-) ).

I have also added some integration tests in a last commit that show the current behaviour in some strange scenarios. After working on this so much time I no longer know what works as expected or not... so I just document it and I let others decide for me :-P And if the tests themselves do not make sense either just drop the last commit; I sense a bug (not from this pull request, but coming already from the previous code) lurking around in the code that returns a reduced set of shares when there are not enough resharing rights, but as much as I tried I could not find it. So maybe everything is just fine :-P

skjnldsv commented:
> so if you changed it for some specific reason please provide a test that shows it

Argh, I cannot recall! :(
I think aside from cleaning the code and not showing your own shares in the getShares function, that was all this pr was about.

The issue I noticed was on the new sidebar, when listing the shares of a shared-with-me folder, I could also see the share that made the folder available to me.
Not sure if that's really clear.
But to sums it up, if userA share a folder with resharing rights to userB, and then as userB I open the folder, I see all the shares that userA shared with others on this folder. Since I'm userB, I should not see the initial userA->userB share, I can still unshare it from the files list (instead of delte, that's how we do it).



> I have added more unit and integration tests, split the fixes and the refactorings, did some minor cleanups and fixed the pending issues (I hope :-P ).

That's is an incredible work!!
Thanks a lot Daniel!!! :hugs: :heart:

skjnldsv commented:
> After working on this so much time I no longer know what works as expected or not

that is the issue here.
That's why I split my 4 previous pr into 4 because it's so complicated! :O



> did some minor cleanups and fixed the pending issues (I hope :-P ).

So far:
```
--- Failed scenarios:
699 |  
700 | /drone/src/build/integration/sharing_features/sharing-v1.feature:308
701

```

skjnldsv commented:
Failed integration test now passes.
Acceptances failures are timeouts and were passing on the previous one
```
[Composer\Downloader\TransportException]
The "https://packagist.org/p/provider-2015%24388aba92c0dcf5c49c5ef6281aee51336f0df71b3d6a4111efbe8558d462bb03.json" file could not be downloaded (HTTP/1.1 404 Not Found)
```

:rocket:
